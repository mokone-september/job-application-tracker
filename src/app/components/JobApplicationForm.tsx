'use client';

import { useState, useEffect } from 'react';
import type { JobApplication, ApplicationStatus } from '../types/application';

interface JobApplicationFormProps {
  onSubmit: (application: Omit<JobApplication, 'id'>) => void;
  initialData?: JobApplication;
}

const defaultFormData: Omit<JobApplication, 'id'> = {
  company: '',
  position: '',
  status: 'Applied',
  appliedDate: new Date().toISOString().split('T')[0],
  notes: ''
};

export default function JobApplicationForm({ onSubmit, initialData }: JobApplicationFormProps) {
  const [formData, setFormData] = useState<Omit<JobApplication, 'id'>>(defaultFormData);

  useEffect(() => {
    if (initialData) {
      const { ...rest } = initialData;
      setFormData(rest);
    }
  }, [initialData]);

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    onSubmit(formData);
    if (!initialData) {
      setFormData(defaultFormData);
    }
  };

  return (
    <form 
      onSubmit={handleSubmit} 
      className="bg-white p-6 rounded-lg shadow-md border border-gray-200"
    >
      <h2 className="text-xl font-semibold text-gray-800 mb-6">
        {initialData ? 'Edit Application' : 'Add New Application'}
      </h2>

      <div className="grid grid-cols-1 gap-6 sm:grid-cols-2">
        {/* Company Field */}
        <div>
          <label htmlFor="company" className="block text-sm font-medium text-gray-700 mb-1">
            Company
          </label>
          <input
            id="company"
            type="text"
            value={formData.company}
            onChange={(e) => setFormData({ ...formData, company: e.target.value })}
            className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
            placeholder="Enter company name"
            required
          />
        </div>

        {/* Position Field */}
        <div>
          <label htmlFor="position" className="block text-sm font-medium text-gray-700 mb-1">
            Position
          </label>
          <input
            id="position"
            type="text"
            value={formData.position}
            onChange={(e) => setFormData({ ...formData, position: e.target.value })}
            className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
            placeholder="Enter position title"
            required
          />
        </div>

        {/* Status Field */}
        <div>
          <label htmlFor="status" className="block text-sm font-medium text-gray-700 mb-1">
            Status
          </label>
          <select
            id="status"
            value={formData.status}
            onChange={(e) => setFormData({ ...formData, status: e.target.value as ApplicationStatus })}
            className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          >
            <option value="Applied">Applied</option>
            <option value="Interviewing">Interviewing</option>
            <option value="Offered">Offered</option>
            <option value="Rejected">Rejected</option>
          </select>
        </div>

        {/* Applied Date Field */}
        <div>
          <label htmlFor="appliedDate" className="block text-sm font-medium text-gray-700 mb-1">
            Applied Date
          </label>
          <input
            id="appliedDate"
            type="date"
            value={formData.appliedDate}
            onChange={(e) => setFormData({ ...formData, appliedDate: e.target.value })}
            className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
            required
          />
        </div>
      </div>

      {/* Notes Field */}
      <div className="mt-6">
        <label htmlFor="notes" className="block text-sm font-medium text-gray-700 mb-1">
          Notes
        </label>
        <textarea
          id="notes"
          value={formData.notes}
          onChange={(e) => setFormData({ ...formData, notes: e.target.value })}
          className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          rows={4}
          placeholder="Add any additional notes"
        />
      </div>

      {/* Submit Button */}
      <div className="mt-6">
        <button
          type="submit"
          className="w-full sm:w-auto px-6 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
        >
          {initialData ? 'Update Application' : 'Add Application'}
        </button>
      </div>
    </form>
  );
}