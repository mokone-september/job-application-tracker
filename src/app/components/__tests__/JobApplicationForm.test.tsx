import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import JobApplicationForm from '../JobApplicationForm';
import type { JobApplication, ApplicationStatus } from '../../types/application';

describe('JobApplicationForm', () => {
  const mockSubmit = jest.fn();

  const mockInitialData: JobApplication = {
    id: 1,
    company: 'Tech Corp',
    position: 'Frontend Developer',
    status: 'Interviewing' as ApplicationStatus,
    appliedDate: '2023-05-15',
    notes: 'Second round interview scheduled'
  };

  beforeEach(() => {
    mockSubmit.mockClear();
  });

  it('renders the form with default values when no initial data', () => {
    render(<JobApplicationForm onSubmit={mockSubmit} />);
    
    expect(screen.getByLabelText('Company')).toHaveValue('');
    expect(screen.getByLabelText('Position')).toHaveValue('');
    expect(screen.getByLabelText('Status')).toHaveValue('Applied');
    expect(screen.getByLabelText('Applied Date')).toHaveValue(new Date().toISOString().split('T')[0]);
    expect(screen.getByLabelText('Notes')).toHaveValue('');
    expect(screen.getByRole('button')).toHaveTextContent('Add Application');
  });

  it('renders with initial data when provided', () => {
    render(<JobApplicationForm onSubmit={mockSubmit} initialData={mockInitialData} />);
    
    expect(screen.getByLabelText('Company')).toHaveValue(mockInitialData.company);
    expect(screen.getByLabelText('Position')).toHaveValue(mockInitialData.position);
    expect(screen.getByLabelText('Status')).toHaveValue(mockInitialData.status);
    expect(screen.getByLabelText('Applied Date')).toHaveValue(mockInitialData.appliedDate);
    expect(screen.getByLabelText('Notes')).toHaveValue(mockInitialData.notes);
    expect(screen.getByRole('button')).toHaveTextContent('Update Application');
  });

  it('submits form data when valid', async () => {
    render(<JobApplicationForm onSubmit={mockSubmit} />);
    
    await userEvent.type(screen.getByLabelText('Company'), 'Google');
    await userEvent.type(screen.getByLabelText('Position'), 'Software Engineer');
    await userEvent.selectOptions(screen.getByLabelText('Status'), 'Interviewing');
    await userEvent.type(screen.getByLabelText('Applied Date'), '2023-06-01');
    await userEvent.type(screen.getByLabelText('Notes'), 'Referral from John Doe');
    
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(mockSubmit).toHaveBeenCalledWith({
        company: 'Google',
        position: 'Software Engineer',
        status: 'Interviewing',
        appliedDate: '2023-06-01',
        notes: 'Referral from John Doe'
      });
    });
  });

  it('resets form after submission when no initial data', async () => {
    render(<JobApplicationForm onSubmit={mockSubmit} />);
    
    await userEvent.type(screen.getByLabelText('Company'), 'Amazon');
    await userEvent.type(screen.getByLabelText('Position'), 'Backend Developer');
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(screen.getByLabelText('Company')).toHaveValue('');
      expect(screen.getByLabelText('Position')).toHaveValue('');
    });
  });

  it('does not reset form after submission when editing', async () => {
    render(<JobApplicationForm onSubmit={mockSubmit} initialData={mockInitialData} />);
    
    const newCompany = 'Updated Company';
    await userEvent.clear(screen.getByLabelText('Company'));
    await userEvent.type(screen.getByLabelText('Company'), newCompany);
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(screen.getByLabelText('Company')).toHaveValue(newCompany);
    });
  });

  it('validates required fields', async () => {
    render(<JobApplicationForm onSubmit={mockSubmit} />);
    
    fireEvent.click(screen.getByRole('button'));
    
    await waitFor(() => {
      expect(mockSubmit).not.toHaveBeenCalled();
    });
    
    expect(screen.getByLabelText('Company')).toHaveAttribute('required');
    expect(screen.getByLabelText('Position')).toHaveAttribute('required');
    expect(screen.getByLabelText('Applied Date')).toHaveAttribute('required');
  });

  it('handles status dropdown changes', async () => {
    render(<JobApplicationForm onSubmit={mockSubmit} />);
    
    const statusSelect = screen.getByLabelText('Status');
    await userEvent.selectOptions(statusSelect, 'Rejected');
    
    expect(statusSelect).toHaveValue('Rejected');
  });

  it('matches snapshot when empty', () => {
    const { asFragment } = render(<JobApplicationForm onSubmit={mockSubmit} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with initial data', () => {
    const { asFragment } = render(
      <JobApplicationForm onSubmit={mockSubmit} initialData={mockInitialData} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

