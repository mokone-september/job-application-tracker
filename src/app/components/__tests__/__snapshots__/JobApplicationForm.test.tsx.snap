// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobApplicationForm matches snapshot when empty 1`] = `
<DocumentFragment>
  <form
    class="bg-white p-6 rounded-lg shadow-md border border-gray-200"
  >
    <h2
      class="text-xl font-semibold text-gray-800 mb-6"
    >
      Add New Application
    </h2>
    <div
      class="grid grid-cols-1 gap-6 sm:grid-cols-2"
    >
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="company"
        >
          Company
        </label>
        <input
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="company"
          placeholder="Enter company name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="position"
        >
          Position
        </label>
        <input
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="position"
          placeholder="Enter position title"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="status"
        >
          Status
        </label>
        <select
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="status"
        >
          <option
            value="Applied"
          >
            Applied
          </option>
          <option
            value="Interviewing"
          >
            Interviewing
          </option>
          <option
            value="Offered"
          >
            Offered
          </option>
          <option
            value="Rejected"
          >
            Rejected
          </option>
        </select>
      </div>
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="appliedDate"
        >
          Applied Date
        </label>
        <input
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="appliedDate"
          required=""
          type="date"
          value="2025-04-05"
        />
      </div>
    </div>
    <div
      class="mt-6"
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-1"
        for="notes"
      >
        Notes
      </label>
      <textarea
        class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
        id="notes"
        placeholder="Add any additional notes"
        rows="4"
      />
    </div>
    <div
      class="mt-6"
    >
      <button
        class="w-full sm:w-auto px-6 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
        type="submit"
      >
        Add Application
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`JobApplicationForm matches snapshot with initial data 1`] = `
<DocumentFragment>
  <form
    class="bg-white p-6 rounded-lg shadow-md border border-gray-200"
  >
    <h2
      class="text-xl font-semibold text-gray-800 mb-6"
    >
      Edit Application
    </h2>
    <div
      class="grid grid-cols-1 gap-6 sm:grid-cols-2"
    >
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="company"
        >
          Company
        </label>
        <input
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="company"
          placeholder="Enter company name"
          required=""
          type="text"
          value="Tech Corp"
        />
      </div>
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="position"
        >
          Position
        </label>
        <input
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="position"
          placeholder="Enter position title"
          required=""
          type="text"
          value="Frontend Developer"
        />
      </div>
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="status"
        >
          Status
        </label>
        <select
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="status"
        >
          <option
            value="Applied"
          >
            Applied
          </option>
          <option
            value="Interviewing"
          >
            Interviewing
          </option>
          <option
            value="Offered"
          >
            Offered
          </option>
          <option
            value="Rejected"
          >
            Rejected
          </option>
        </select>
      </div>
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="appliedDate"
        >
          Applied Date
        </label>
        <input
          class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="appliedDate"
          required=""
          type="date"
          value="2023-05-15"
        />
      </div>
    </div>
    <div
      class="mt-6"
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-1"
        for="notes"
      >
        Notes
      </label>
      <textarea
        class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
        id="notes"
        placeholder="Add any additional notes"
        rows="4"
      >
        Second round interview scheduled
      </textarea>
    </div>
    <div
      class="mt-6"
    >
      <button
        class="w-full sm:w-auto px-6 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
        type="submit"
      >
        Update Application
      </button>
    </div>
  </form>
</DocumentFragment>
`;
